File: app/src/main/java/com/elementary/tasks/core/services/BirthdayActionService.java
package com.elementary.tasks.core.services;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import androidx.annotation.Nullable;

import com.elementary.tasks.Actions;
import com.elementary.tasks.birthdays.BirthdayItem;
import com.elementary.tasks.birthdays.ShowBirthdayActivity;
import com.elementary.tasks.core.app_widgets.UpdatesHelper;
import com.elementary.tasks.core.utils.Constants;
import com.elementary.tasks.core.utils.LogUtil;
import com.elementary.tasks.core.utils.Notifier;
import com.elementary.tasks.core.utils.Permissions;
import com.elementary.tasks.core.utils.RealmDb;
import com.elementary.tasks.core.utils.TelephonyUtil;

import java.util.Calendar;

/**
 * Copyright 2017 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
public class BirthdayActionService extends BroadcastReceiver {

    public static final String ACTION_SHOW = Actions.Birthday.ACTION_SHOW_FULL;
    public static final String ACTION_CALL = Actions.Birthday.ACTION_CALL;
    public static final String ACTION_SMS = Actions.Birthday.ACTION_SMS;

    private static final String TAG = "BirthdayActionService";

    public static Intent hide(Context context, String id) {
        return intent(context, id, PermanentBirthdayReceiver.ACTION_HIDE);
    }

    public static Intent call(Context context, String id) {
        return intent(context, id, ACTION_CALL);
    }

    public static Intent show(Context context, String id) {
        return intent(context, id, ACTION_SHOW);
    }

    public static Intent sms(Context context, String id) {
        return intent(context, id, ACTION_SMS);
    }

    private static Intent intent(Context context, String id, String action) {
        Intent intent = new Intent(context, BirthdayActionService.class);
        intent.setAction(action);
        intent.putExtra(Constants.INTENT_ID, id);
        return intent;
    }

    private void updateBirthday(BirthdayItem item) {
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(System.currentTimeMillis());
        int year = calendar.get(Calendar.YEAR);
        item.setShowedYear(year);
        RealmDb.getInstance().saveObject(item);
    }

    private void sendSms(Context context, Intent intent) {
        BirthdayItem item = RealmDb.getInstance().getBirthday(intent.getStringExtra(Constants.INTENT_ID));
        if (item != null && Permissions.checkPermission(context, Permissions.SEND_SMS)) {
            TelephonyUtil.sendSms(item.getNumber(), context);
            updateBirthday(item);
            finish(context, item.getUniqueId());
        } else {
            hidePermanent(context, intent.getStringExtra(Constants.INTENT_ID));
        }
    }

    private void makeCall(Context context, Intent intent) {
        BirthdayItem item = RealmDb.getInstance().getBirthday(intent.getStringExtra(Constants.INTENT_ID));
        if (item != null && Permissions.checkPermission(context, Permissions.CALL_PHONE)) {
            TelephonyUtil.makeCall(item.getNumber(), context);
            updateBirthday(item);
            finish(context, item.getUniqueId());
        } else {
            hidePermanent(context, intent.getStringExtra(Constants.INTENT_ID));
        }
    }

    private void showReminder(Context context, Intent intent) {
        BirthdayItem reminder = RealmDb.getInstance().getBirthday(intent.getStringExtra(Constants.INTENT_ID));
        if (reminder != null) {
            Intent notificationIntent = ShowBirthdayActivity.getLaunchIntent(context,
                    intent.getStringExtra(Constants.INTENT_ID));
            notificationIntent.putExtra(Constants.INTENT_NOTIFICATION, true);
            context.startActivity(notificationIntent);
            Notifier.hideNotification(context, PermanentBirthdayReceiver.BIRTHDAY_PERM_ID);
        }
    }

    private void hidePermanent(Context context, @Nullable String id) {
        if (id == null) return;
        BirthdayItem item = RealmDb.getInstance().getBirthday(id);
        if (item != null) {
            updateBirthday(item);
            finish(context, item.getUniqueId());
        }
    }

    private static void finish(Context context, int id) {
        Notifier.hideNotification(context, id);
        UpdatesHelper.getInstance(context).updateWidget();
        UpdatesHelper.getInstance(context).updateCalendarWidget();
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        if (intent != null) {
            String action = intent.getAction();
            LogUtil.d(TAG, "onStartCommand: " + action);
            if (action != null) {
                if (action.matches(ACTION_CALL)) {
                    makeCall(context, intent);
                } else if (action.matches(ACTION_SMS)) {
                    sendSms(context, intent);
                } else if (action.matches(PermanentBirthdayReceiver.ACTION_HIDE)) {
                    hidePermanent(context, intent.getStringExtra(Constants.INTENT_ID));
                } else {
                    showReminder(context, intent);
                }
            }
        }
    }
}

File: app/src/main/java/com/elementary/tasks/core/services/BirthdayActionService.kt
package com.elementary.tasks.core.services

import android.content.Context
import android.content.Intent
import com.elementary.tasks.Actions
import com.elementary.tasks.ReminderApp
import com.elementary.tasks.birthdays.ShowBirthdayActivity
import com.elementary.tasks.core.appWidgets.UpdatesHelper
import com.elementary.tasks.core.data.AppDb
import com.elementary.tasks.core.data.models.Birthday
import com.elementary.tasks.core.utils.*
import java.util.*
import javax.inject.Inject

/**
 * Copyright 2017 Nazar Suhovich
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
class BirthdayActionService : BaseBroadcast() {

    @Inject
    lateinit var updatesHelper: UpdatesHelper

    init {
        ReminderApp.appComponent.inject(this)
    }

    private fun updateBirthday(context: Context, item: Birthday) {
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = System.currentTimeMillis()
        val year = calendar.get(Calendar.YEAR)
        item.showedYear = year
        AppDb.getAppDatabase(context).birthdaysDao().insert(item)
    }

    private fun sendSms(context: Context, intent: Intent) {
        val item = AppDb.getAppDatabase(context).birthdaysDao().getById(intent.getIntExtra(Constants.INTENT_ID, 0))
        if (item != null && Permissions.checkPermission(context, Permissions.SEND_SMS)) {
            TelephonyUtil.sendSms(item.number, context)
            updateBirthday(context, item)
            finish(notifier, updatesHelper, item.uniqueId)
        } else {
            hidePermanent(context, intent.getIntExtra(Constants.INTENT_ID, 0))
        }
    }

    private fun makeCall(context: Context, intent: Intent) {
        val item = AppDb.getAppDatabase(context).birthdaysDao().getById(intent.getIntExtra(Constants.INTENT_ID, 0))
        if (item != null && Permissions.checkPermission(context, Permissions.CALL_PHONE)) {
            TelephonyUtil.makeCall(item.number, context)
            updateBirthday(context, item)
            finish(notifier, updatesHelper, item.uniqueId)
        } else {
            hidePermanent(context, intent.getIntExtra(Constants.INTENT_ID, 0))
        }
    }

    private fun showReminder(context: Context, intent: Intent) {
        val reminder = AppDb.getAppDatabase(context).birthdaysDao().getById(intent.getIntExtra(Constants.INTENT_ID, 0))
        if (reminder != null) {
            val notificationIntent = ShowBirthdayActivity.getLaunchIntent(context,
                    intent.getIntExtra(Constants.INTENT_ID, 0))
            notificationIntent.putExtra(Constants.INTENT_NOTIFICATION, true)
            context.startActivity(notificationIntent)
            notifier.hideNotification(PermanentBirthdayReceiver.BIRTHDAY_PERM_ID)
        }
    }

    private fun hidePermanent(context: Context, id: Int) {
        if (id == 0) return
        val item = AppDb.getAppDatabase(context).birthdaysDao().getById(id)
        if (item != null) {
            updateBirthday(context, item)
            finish(notifier, updatesHelper, item.uniqueId)
        }
    }

    override fun onReceive(context: Context, intent: Intent?) {
        if (intent != null) {
            val action = intent.action
            LogUtil.d(TAG, "onStartCommand: " + action!!)
            when {
                action.matches(ACTION_CALL.toRegex()) -> makeCall(context, intent)
                action.matches(ACTION_SMS.toRegex()) -> sendSms(context, intent)
                action.matches(PermanentBirthdayReceiver.ACTION_HIDE.toRegex()) -> hidePermanent(context, intent.getIntExtra(Constants.INTENT_ID, 0))
                else -> showReminder(context, intent)
            }
        }
    }

    companion object {

        const val ACTION_SHOW = Actions.Birthday.ACTION_SHOW_FULL
        const val ACTION_CALL = Actions.Birthday.ACTION_CALL
        const val ACTION_SMS = Actions.Birthday.ACTION_SMS

        private const val TAG = "BirthdayActionService"

        fun hide(context: Context, id: Int): Intent {
            return intent(context, id, PermanentBirthdayReceiver.ACTION_HIDE)
        }

        fun call(context: Context, id: Int): Intent {
            return intent(context, id, ACTION_CALL)
        }

        fun show(context: Context, id: Int): Intent {
            return intent(context, id, ACTION_SHOW)
        }

        fun sms(context: Context, id: Int): Intent {
            return intent(context, id, ACTION_SMS)
        }

        private fun intent(context: Context, id: Int, action: String): Intent {
            val intent = Intent(context, BirthdayActionService::class.java)
            intent.action = action
            intent.putExtra(Constants.INTENT_ID, id)
            return intent
        }

        private fun finish(notifier: Notifier, updatesHelper: UpdatesHelper, id: Int) {
            notifier.hideNotification( id)
            updatesHelper.updateWidget()
            updatesHelper.updateCalendarWidget()
        }
    }
}
